File: euler-problems/problems1-10/problem7/main.ex
defmodule Problem7 do
  # By listing the first six prime numbers: 2, 3, 5, 7, 11, and 13, we can see that the 6th prime is 13.
  # What is the 10 001st prime number?
  # Result: 104743

  require Float

  def divisible_by_any(n, list) do
    list |> Enum.any?(&(rem(n, &1) == 0))
  end

  def solve2 do
    n = 10001

    Stream.iterate(1, &(&1 + 2))
    |> Stream.scan([2], fn
      e, acc when e <= 2 -> acc
      e, acc -> if divisible_by_any(e, acc), do: acc, else: [e | acc]
    end)
    |> Stream.map(&List.first/1)
    |> Stream.dedup()
    |> Stream.take(n)
    |> Enum.at(-1)
  end

  def is_prime(n), do: !divisible_by_any(n, 2..(:math.sqrt(n) |> Float.floor() |> round()))

  def solve1 do
    n = 10001

    Stream.concat(
      [2, 3],
      Stream.iterate(5, &(&1 + 1))
      |> Stream.filter(&is_prime/1)
    )
    |> Stream.take(n)
    |> Enum.at(-1)
  end

  def main do
    IO.puts(solve1())
    IO.puts(solve2())
  end
end

File: euler-problems/problems1-10/problem7/main.dart
import 'dart:math' show sqrt;

/**
 * By listing the first six prime numbers: 2, 3, 5, 7, 11, and 13, we can see that the 6th prime is 13.
 * What is the 10 001st prime number?
 * Result: 104743
 */

int solve1() {
  int n_th = 10001;
  int i = 0;
  int n = 2;

  while (true) {
    if (is_prime(n)) {
      ++i;
      if (i == n_th) {
        return n;
      }
    }
    ++n;
  }
}

bool is_prime(int n) =>
    n >= 2 &&
    new Iterable.generate(sqrt(n).floor() + 1).skip(2).every((i) => n % i != 0);

Iterable<int> get primes sync* {
  int i = 2;
  while (true) {
    if (is_prime(i)) {
      yield i;
    }
    ++i;
  }
}

int solve2() {
  int n_th = 10001;
  return primes.elementAt(n_th - 1);
}

main(List<String> arguments) {
  print(solve1());
  print(solve2());
}
